### PR TITLE
make lifetimes that only appear in return type early-bound

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is the full and proper fix for #32330. This also makes some effort to give a nice error message (as evidenced by the `ui` test), sending users over to the tracking issue for a fuller explanation and offering a `--explain` message in some cases.

This needs a crater run before we land.

r? @arielb1

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Oh I forgot-- I should deprecate the older fwds-compat lint too. And maybe improve wording of error message? Not sure how to do that.

**Comment by nikomatsakis** [comment]
Crater run results: https://gist.github.com/nikomatsakis/25076c557d1c7d5ce3bcbacffdc953c9

### Root regressions, sorted by rank:

* [ ] [lmdb-rs-0.7.3](https://crates.io/crates/lmdb-rs) ([before](https://tools.taskcluster.net/task-inspector/#KDLRhhxaRC2ESWOYvhAAfw)) ([after](https://tools.taskcluster.net/task-inspector/#vrxtieThRoGNzdMAPBiK-A))
* [ ] [rusted_cypher-0.9.1](https://crates.io/crates/rusted_cypher) ([before](https://tools.taskcluster.net/task-inspector/#HYOV8WHrRumTsc4S5QQg3g)) ([after](https://tools.taskcluster.net/task-inspector/#ol5oGtsKRemGm34dEyiMag))
* [ ] [doubly-1.1.1](https://crates.io/crates/doubly) ([before](https://tools.taskcluster.net/task-inspector/#1DDahewPQbCMqIRhi0iSPA)) ([after](https://tools.taskcluster.net/task-inspector/#E7k1FB3PRoiFHhoIqeWG8Q))
* [ ] [nametable_codegen-0.1.2](https://crates.io/crates/nametable_codegen) ([before](https://tools.taskcluster.net/task-inspector/#ztjMxbrtSS6g2UEp4M4pyw)) ([after](https://tools.taskcluster.net/task-inspector/#PNIh-2BqThyCLo4d2cRuhQ))
* [ ] [pandoc_ast-0.4.0](https://crates.io/crates/pandoc_ast) ([before](https://tools.taskcluster.net/task-inspector/#gYDxBUPoTFOfuPdH_EtE3A)) ([after](https://tools.taskcluster.net/task-inspector/#g30UJU8jQ1SmDVEtqF0AxA))
* [ ] [parity-dapps-glue-1.4.0](https://crates.io/crates/parity-dapps-glue) ([before](https://tools.taskcluster.net/task-inspector/#uVXbXJ9iSxW3HKt7hXV8VA)) ([after](https://tools.taskcluster.net/task-inspector/#ieCzVlQQQOiCsA6DYlK5-A))
* [ ] [pippin-0.1.0](https://crates.io/crates/pippin) ([before](https://tools.taskcluster.net/task-inspector/#7-m0qrlKQhGAFl-GbJD6qg)) ([after](https://tools.taskcluster.net/task-inspector/#89Q4inFtSr6KeSqNO1EYHw))
* [ ] [presto-0.1.0](https://crates.io/crates/presto) ([before](https://tools.taskcluster.net/task-inspector/#jCuehWI_Qk-CYCrUfRhKgQ)) ([after](https://tools.taskcluster.net/task-inspector/#DGJRvgjEQbKy0DKxb2G7tw))
* [ ] [presto-rs-0.1.0](https://crates.io/crates/presto-rs) ([before](https://tools.taskcluster.net/task-inspector/#-Ugl_2bXQ4eJmyEfpL-uQw)) ([after](https://tools.taskcluster.net/task-inspector/#g2xbLOnKRCST8YYSzMra6A))
* [ ] [tiberius-0.1.0](https://crates.io/crates/tiberius) ([before](https://tools.taskcluster.net/task-inspector/#gGEueUHoRyK2HmUYGujdMA)) ([after](https://tools.taskcluster.net/task-inspector/#3rONwTmXQDaLsdmbj1O9Rg))

**Comment by nikomatsakis** [comment]
Some of the failures above may be addressed by fixing #33684, though I'm not sure. In any case, I can try to submit PRs to address the issues.

**Comment by arielb1** [comment]
Will review later today. Sorry I missed this.

File: src/librustc/middle/resolve_lifetime.rs [hunk 1]
```diff
@@ -54,10 +54,14 @@ pub struct NamedRegionMap {
     // `DefRegion` describing how that region is bound
     pub defs: NodeMap<DefRegion>,
 
-    // the set of lifetime def ids that are late-bound; late-bound ids
-    // are named regions appearing in fn arguments that do not appear
-    // in where-clauses
-    pub late_bound: NodeMap<ty::Issue32330>,
+    // the set of lifetime def ids that are early-bound; early-bound
```
Comment (arielb1) [inline src/librustc/middle/resolve_lifetime.rs]: Should be "that are late-bound".

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -637,8 +638,7 @@ impl<'tcx> RegionParameterDef<'tcx> {
     }
 
     pub fn to_bound_region(&self) -> ty::BoundRegion {
-        // this is an early bound region, so unaffected by #32330
-        ty::BoundRegion::BrNamed(self.def_id, self.name, Issue32330::WontChange)
+        ty::BoundRegion::BrNamed(self.def_id, self.name, self.issue_32330)
```
Comment (arielb1) [inline src/librustc/ty/mod.rs]: So we're having that big enum on every `ty::Region`? I suppose we should intern the `(DefId, Name, Issue32230)` triple and check the perf impact.

**Review by arielb1 — CHANGES_REQUESTED** [review]
r+ after we communicate with the crater crate authors and fix nits.

**Comment by arielb1** [comment]
```
tidy check (x86_64-unknown-linux-gnu)

duplicate error code: 573

/checkout/src/librustc_typeck/diagnostics.rs:4138:     E0573: r##"

/checkout/src/librustc_resolve/diagnostics.rs:1544:     E0573,

duplicate error code: 574

/checkout/src/librustc_typeck/diagnostics.rs:4175:     E0574: r##"

/checkout/src/librustc_resolve/diagnostics.rs:1545:     E0574,

thread 'main' panicked at 'some tidy checks failed', /checkout/src/tools/tidy/src/main.rs:55

note: Run with `RUST_BACKTRACE=1` for a backtrace.

```

**Comment by nikomatsakis** [comment]
@arielb1 ok I will follow-up prob tomorrow...

**Comment by nikomatsakis** [comment]
OK, this PR is ready to land. The following are the "core crates" that appear to be affected by this change:

- [ ] lmdb-rs (cc @vhbit)
- [ ] doubly (cc @marchelzo)
- [ ] pippin (cc @dhardy)

These crates will have to be fixed -- hopefully they are getting a lint warning now about future compatibility, though I'm not sure if the lint covers *all* the cases that are affected by the proper fix (that is quite challenging). Let me know if you are having trouble updating -- if I find time, I will try to prep some PRs. =)

In addition, a number of crates were affected indirectly, because they relied on an older serde release that is affected. In these cases, the ideal thing is that they would update to a modern version of serde:

- [ ] `rusted-cypher` (cc @livioribeiro)
- [ ] `nametable` (cc @crlf0710)
- [ ] `pandoc-ast` (cc @oli-obk)
- [ ] `parity-dapps-glue` (cc @ethcore)
- [ ] `presto` and `presto-rs` (cc @ernestrc, but these appear to be unavailable in GitHub)

We could also solve this by a new point release of the relevant serde crates (cc @dtolnay, @erickt).

**Comment by nikomatsakis** [comment]
@arielb1 

> So we're having that big enum on every ty::Region? I suppose we should intern the (DefId, Name, Issue32230) triple and check the perf impact. Maybe just do bitpacking to fit BoundRegion in a u32 - should check the perf impact of that.

Ah, I missed this comment. Yes, that's a reasonable point. I'll take a look and see what I can do.

**Comment by nikomatsakis** [comment]
@arielb1 

> So we're having that big enum on every ty::Region? I suppose we should intern the (DefId, Name, Issue32230) triple and check the perf impact. Maybe just do bitpacking to fit BoundRegion in a u32 - should check the perf impact of that.

Fixed in latest commit. PR should be good to go.

**Comment by eddyb** [comment]
* cycling Travis to force a build from it

**Comment by nikomatsakis** [comment]
Tagging this for release notes. Something like this:

"Lifetime parameters that do not appear in the arguments are now considered early-bound, resolving a soundness bug (#32330). The `hr_lifetime_in_assoc_type` future-compatibility lint has been in effect since 1.8.0."

**Pull request merged**